akka-stream/src/main/scala/akka/stream/impl/FastDroppingQueue.scala

```diff
@@ -0,0 +1,182 @@
+    case object Running extends State
+    case class Done(result: OfferResult.CompletionResult) extends State
+
+    val state = new AtomicReference[State](Running)
+
+    val queue = new AbstractBoundedNodeQueue[T](bufferSize) {}
+
+    object Logic extends GraphStageLogic(shape) with OutHandler {
+      setHandler(out, this)
+      val callback = getAsyncCallback[Unit] { _ =>
+        clearNeedsActivation()
+        run()
+      }
+
+      override def onPull(): Unit = run()
+
+      override def onDownstreamFinish(): Unit = {
+        setDone(Done(OfferResult.Cancelled))
+
+        super.onDownstreamFinish()
+      }
+
+      override def postStop(): Unit =
+        // drain queue
+        while (!queue.isEmpty) queue.poll()
+
+      /**
+       * Main loop of the queue. We do two volatile reads for the fast path of pushing elements
+       * from the queue to the stream: one for the state and one to poll the queue. This leads to a somewhat simple design
+       * that will quickly pick up failures from the queue interface.
+       *
+       * An even more optimized version could use a fast path in onPull to avoid reading the state for every element.
+       */
+      @tailrec
+      def run(): Unit =
+        state.get() match {
+          case Running =>
+            if (isAvailable(out)) {
+              val next = queue.poll()
+              if (next == null) { // queue empty
+                if (!setNeedsActivation())
+                  run() // didn't manage to set because stream has been completed in the meantime
+                else if (!queue.isEmpty) /* && setNeedsActivation was true */ {
+                  // tricky case: new element might have been added in the meantime without callback being sent because
+                  // NeedsActivation had not yet been set
+
+                  clearNeedsActivation()
+                  run()
+                } // else Queue.isEmpty && setNeedsActivation was true: waiting for next offer
+              } else
+                push(out, next) // and then: wait for pull
+            } // else: wait for pull
+
+          case Done(OfferResult.Completed) =>
+            if (queue.isEmpty) completeStage()
+            else if (isAvailable(out)) {
+              push(out, queue.poll())
+              run() // another round, might be empty now
+            }
+          // else !Queue.isEmpty: wait for pull to drain remaining elements
+          case Done(OfferResult.Failed(ex)) => failStage(ex)
+          case Done(OfferResult.Cancelled)  => throw new IllegalStateException // should not happen
+          case NeedsActivation              => throw new IllegalStateException // needs to be cleared before
+        }
+    }
+
+    object Mat extends FastDroppingQueue[T] {
+      override def offer(elem: T): OfferResult = state.get() match {
+        case Done(result) => result
+        case x =>
+          if (queue.add(elem)) {
+            if (x == NeedsActivation)
```

Comment:
The bug is here, `state` needs to be read (again?) after `queue.add` to query if activation is needed.